### PR TITLE
implement download + upload s3

### PR DESCRIPTION
**Pull request by apelstein**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -2,7 +2,59 @@
...
+  source_path_in_s3 = 'my-bucket-dir-2/'  # when downloading a dir, path should end with '/'
+  target_path = './my-local-dir-2/'       # when downloading a dir, path should end with '/'
+  regex_filter = 'file_prefix_[0-9][0-9][0-9]' # file_prefix_100, file_prefix_123, file_prefix_754... 
+  
+  download_from_s3(source_path=source_path_in_s3, target_path=target_path, regex_filter=regex_filter)
+      
+  ```
+  #### Upload to s3:
+  ```
+  # example usage:
```
Comment (daTokenizer) [inline README.md:53]:
```suggestion
  #### Upload a filtered set of files from a *local* directory to s3:
  ```
```

File: README.md [hunk 2]
```diff
@@ -2,7 +2,84 @@
...
+First, set your 3 secrets: bucket name, access key and secret key as environment variables. 
+Can be done anyway, for example with python:
+  ```python
+import os
+
+os.environ['HYPERPLANE_SECRET_s3_bucket_name']='my_aws_bucket_name'
+os.environ['HYPERPLANE_SECRET_s3_access_key_secret']='my_aws_access_key_secret'
+os.environ['HYPERPLANE_SECRET_s3_access_key_id']='my_aws_access_key_id'
+  ```
+  #### Download a file from s3:
```
Comment (daTokenizer) [inline README.md:21]:
```suggestion
  #### Download a single file from s3 to your local machine:
```

File: README.md [hunk 3]
```diff
@@ -2,7 +2,84 @@
...
+  ```python
+from hyperplane.aws_s3_utils import download_from_s3
+
+source_path_in_s3 = 'my-bucket-dir-1/file-name-in-s3-1.data'
+target_path = './my-local-dir-1/file-name-in-my-computer-1.data'
+
+download_from_s3(source_path=source_path_in_s3, target_path=target_path)
+      
+  ```
+  #### Upload a file to s3:
```
Comment (daTokenizer) [inline README.md:31]:
```suggestion
  #### Upload a file to s3 from your local machine:
```

File: README.md [hunk 4]
```diff
@@ -2,7 +2,84 @@
...
+  ```python
+from hyperplane.aws_s3_utils import download_from_s3
+
+source_path_in_s3 = 'my-bucket-dir-2/'  # when downloading a dir, path should end with '/'
+target_path = './my-local-dir-2/'       # when downloading a dir, path should end with '/'
+
+download_from_s3(source_path=source_path_in_s3, target_path=target_path)
+      
+  ```
+  #### Upload a filtered set of files from a *local* directory to s3:
```
Comment (daTokenizer) [inline README.md:65]:
```suggestion
  #### Upload a filtered set of files from a *job server* directory to s3:
```

File: README.md [hunk 5]
```diff
@@ -2,7 +2,84 @@
...
+from hyperplane.aws_s3_utils import upload_to_s3
+
+target_path = 'my-bucket-dir-2/'                  # when uploading a dir, path should end with '/'
+source_path_in_my_computer = './my-local-dir-2/'  # when uploading a dir, path should end with '/'
+regex_filter = 'file_prefix_[0-9][0-9][0-9]' # file_prefix_100, file_prefix_123, file_prefix_754... 
+
+file_paths = upload_to_s3(source_path=source_path_in_my_computer, target_path=target_path, regex_filter=regex_filter)
+      
+  ```
+  #### Upload complete directory from a *local* directory to s3:
```
Comment (daTokenizer) [inline README.md:76]:
```suggestion
  #### Upload complete directory from a *job server* directory to s3:
```

File: README.md [hunk 6]
```diff
@@ -2,7 +2,84 @@
...
+  ```
+  #### Upload a filtered set of files from a *local* directory to s3:
+  ```python
+from hyperplane.aws_s3_utils import upload_to_s3
+
+target_path = 'my-bucket-dir-2/'                  # when uploading a dir, path should end with '/'
+source_path_in_my_computer = './my-local-dir-2/'  # when uploading a dir, path should end with '/'
+regex_filter = 'file_prefix_[0-9][0-9][0-9]' # file_prefix_100, file_prefix_123, file_prefix_754... 
+
+file_paths = upload_to_s3(source_path=source_path_in_my_computer, target_path=target_path, regex_filter=regex_filter)
```
Comment (daTokenizer) [inline README.md:73]:
```suggestion
file_paths = upload_to_s3(source_path=source_path_in_my_server, target_path=target_path, regex_filter=regex_filter)
```

File: README.md [hunk 7]
```diff
@@ -2,7 +2,84 @@
...
+
+download_from_s3(source_path=source_path_in_s3, target_path=target_path)
+      
+  ```
+  #### Upload a filtered set of files from a *local* directory to s3:
+  ```python
+from hyperplane.aws_s3_utils import upload_to_s3
+
+target_path = 'my-bucket-dir-2/'                  # when uploading a dir, path should end with '/'
+source_path_in_my_computer = './my-local-dir-2/'  # when uploading a dir, path should end with '/'
```
Comment (daTokenizer) [inline README.md:70]:
```suggestion
source_path_in_my_server = './my-server-dir-2/'  # when uploading a dir, path should end with '/'
```

File: README.md [hunk 8]
```diff
@@ -2,7 +2,84 @@
...
+
+file_paths = upload_to_s3(source_path=source_path_in_my_computer, target_path=target_path, regex_filter=regex_filter)
+      
+  ```
+  #### Upload complete directory from a *local* directory to s3:
+  ```python
+from hyperplane.aws_s3_utils import upload_to_s3
+
+target_path = 'my-bucket-dir-2/'                  # when uploading a dir, path should end with '/'
+source_path_in_my_computer = './my-local-dir-2/'  # when uploading a dir, path should end with '/'
```
Comment (daTokenizer) [inline README.md:81]:
```suggestion
source_path_in_my_server = './my-server-dir-2/'  # when uploading a dir, path should end with '/'
```

File: README.md [hunk 9]
```diff
@@ -2,7 +2,84 @@
...
+
+file_paths = upload_to_s3(source_path=source_path_in_my_computer, target_path=target_path)
```
Comment (daTokenizer) [inline README.md:83]:
```suggestion
file_paths = upload_to_s3(source_path=source_path_in_my_server, target_path=target_path)
```

File: src/hyperplane/aws_s3_utils.py [hunk 10]
```diff
@@ -0,0 +1,200 @@
...
+    :param object_name: S3 object name. If not specified then file_name is used
+    :param extra_args: arg to pass to s3 call
+    :param verbose: print logs if true
+    :param regex_filter: regex compile str to match
+    :param s3_client: s3 client to interact with the bucket
+    :return: s3 url if file was uploaded, else None
+
+    """
+    if regex_filter:
+        regex_filter = re.compile(regex_filter)
```
Comment (daTokenizer) [inline src/hyperplane/aws_s3_utils.py:119]: compile this outside of this function, it makes no sense to compile every time for every file (there could be 100o's of them)

File: src/hyperplane/aws_s3_utils.py [hunk 11]
```diff
@@ -0,0 +1,200 @@
...
+        report("User didn't specify bucket name and have no bucket configurations in our system")
+        return None
+    s3_resource = _connect_to_s3_bucket(bucket_name=bucket_name, access_key_id=user_s3_creds.get("access_key_id"),
+                                        access_key_secret=user_s3_creds.get("secret_access_key"),
+                                        s3_client_or_resource=_s3_resource,
+                                        user_s3_creds=user_s3_creds)
+
+    bucket = s3_resource.Bucket(bucket_name)
+    objects_to_download = list(bucket.objects.filter(Prefix=source_path))
+    regex_compile = re.compile(regex_filter) if regex_filter else None
```
Comment (daTokenizer) [inline src/hyperplane/aws_s3_utils.py:48]: compile this outside of this function, it makes no sense to compile every time for every file (there could be 1000's of them)

File: src/hyperplane/aws_s3_utils.py [hunk 12]
```diff
@@ -0,0 +1,200 @@
...
+        aws_access_key_id=access_key_id,
+        aws_secret_access_key=access_key_secret,
+    )
+
+
+def _connect_to_s3_bucket(bucket_name, user_s3_creds, access_key_secret=None, access_key_id=None,
+                          s3_client_or_resource=_s3_client):
+    print(f"Connecting s3 bucket: {bucket_name}")
+    if access_key_id and access_key_secret:
+        print("Connecting to s3 bucket with given access key id and secret access key")
```
Comment (daTokenizer) [inline src/hyperplane/aws_s3_utils.py]:
```suggestion
```

File: src/hyperplane/aws_s3_utils.py [hunk 13]
```diff
@@ -0,0 +1,200 @@
...
+        return s3_client_or_resource(access_key_id, access_key_secret)
+
+    user_bucket_name = user_s3_creds.get("bucket_url")
+    if bucket_name == user_bucket_name:
+        if not access_key_secret:
+            access_key_secret = user_s3_creds.get("secret_access_key")
+        if not access_key_id:
+            access_key_id = user_s3_creds.get("access_key_id")
+
+    return s3_client_or_resource(access_key_id, access_key_secret)
```
Comment (daTokenizer) [inline src/hyperplane/aws_s3_utils.py]:
```suggestion
    print("Connecting to s3 bucket with stored access key id and secret access key")
    return s3_client_or_resource(access_key_id, access_key_secret)
```

File: src/hyperplane/aws_s3_utils.py [hunk 14]
```diff
@@ -0,0 +1,200 @@
...
+    """
+    user_s3_creds = get_s3_credentials()
+
+    bucket_name = user_s3_creds.get("bucket_url")
+    if not bucket_name:
+        report("User didn't specify bucket name and have no bucket configurations in our system")
+        return None
+    s3_resource = _connect_to_s3_bucket(bucket_name=bucket_name, access_key_id=user_s3_creds.get("access_key_id"),
+                                        access_key_secret=user_s3_creds.get("secret_access_key"),
+                                        s3_client_or_resource=_s3_resource,
```
Comment (daTokenizer) [inline src/hyperplane/aws_s3_utils.py]:
```suggestion
                                        s3_client_or_resource_connector=_s3_resource,
```

**Review by daTokenizer — CHANGES_REQUESTED** [review]

File: src/hyperplane/aws_s3_utils.py (hunk 11 above)
Comment (apelstein) [reply src/hyperplane/aws_s3_utils.py:48]: it is outside from the for loop so...

File: README.md [hunk 15]
```diff
@@ -2,7 +2,84 @@
...
+  ```python
+from hyperplane.aws_s3_utils import download_from_s3
+
+source_path_in_s3 = 'my-bucket-dir-1/file-name-in-s3-1.data'
+target_path = './my-local-dir-1/file-name-in-my-computer-1.data'
+
+download_from_s3(source_path=source_path_in_s3, target_path=target_path)
+      
+  ```
+  #### Upload a file to s3 from your local machine:
```
Comment (daTokenizer) [inline README.md:31]:
```suggestion
  #### Upload a single file to s3 from your local machine:
```

**Review by daTokenizer — DISMISSED** [review]

**Review by daTokenizer — APPROVED** [review]

**Pull request merged**